# Project 2 - Kaggle Competition: Predicting Housing Prices in Ames, Iowa

We continue on from the previous notebook, starting with preprocessing

First, we will do our libraries import and read in the data

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

from sklearn.linear_model import LinearRegression, Lasso, Ridge, LassoCV, RidgeCV
from sklearn.model_selection import cross_val_score, train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn import metrics

In [2]:
pd.set_option('display.max_rows', 3000)
pd.set_option('display.max_columns', 80)

We will set the id column as our index, in line with Kaggle submission format.

In [3]:
train = pd.read_csv('../datasets/cleaned_train.csv', index_col = 'id')
test = pd.read_csv('../datasets/cleaned_test.csv', index_col = 'id')

In [4]:
train.shape

(2040, 63)

In [5]:
test.shape

(879, 62)

In [6]:
train.head()

lot_frontage  lot_area  neighborhood  overall_qual  year_built  \
id                                                                    
109          68.0     13517             1             6        1976   
544          43.0     11492             2             7        1996   
153          68.0      7922             1             5        1953   
318          73.0      9802             3             5        2006   
255          82.0     14235             2             6        1900   

     year_remod_add  mas_vnr_area  bsmt_qual  bsmt_cond  bsmt_exposure  \
id                                                                       
109            2005         289.0          3          3              1   
544            1997         132.0          4          3              1   
153            2007           0.0          3          3              1   
318            2007           0.0          4          3              1   
255            1993           0.0          2          4              1   

     bsmtfin_type_1  bsmtfin_sf_1  heating_qc  electrical  1st_flr_sf  \
id                                                                      
109               6         533.0           5           4         725   
544               6         637.0           5           4         913   
153               6         731.0           3           4        1057   
318               1           0.0           4           4         744   
255               1           0.0           3           4         831   

     2nd_flr_sf  gr_liv_area  bsmt_full_bath  full_bath  half_bath  \
id                                                                   
109         754         1479             0.0          2          1   
544        1209         2122             1.0          2          1   
153           0         1057             1.0          1          0   
318         700         1444             0.0          2          1   
255         614         1445             0.0          2          0   

     kitchen_qual  totrms_abvgrd  fireplaces  garage_yr_blt  garage_finish  \
id                                                                           
109             4              6           0         1976.0              2   
544             4              8           1         1997.0              2   
153             4              5           0         1953.0              1   
318             3              7           0         2007.0              3   
255             3              6           0         1957.0              1   

     garage_cars  garage_area  garage_cond  paved_drive  wood_deck_sf  \
id                                                                      
109          2.0        475.0            3            2             0   
544          2.0        559.0            3            2             0   
153          1.0        246.0            3            2             0   
318          2.0        400.0            3            2           100   
255          2.0        484.0            3            0             0   

     open_porch_sf  saleprice  mas_vnr_type_BrkFace  mas_vnr_type_None  \
id                                                                       
109             44     130500                   1.0                0.0   
544             74     220000                   1.0                0.0   
153             52     109000                   0.0                1.0   
318              0     174000                   0.0                1.0   
255             59     138500                   0.0                1.0   

     mas_vnr_type_Stone  exterior_2nd_AsphShn  exterior_2nd_Brk Cmn  \
id                                                                    
109                 0.0                   0.0                   0.0   
544                 0.0                   0.0                   0.0   
153                 0.0                   0.0                   0.0   
318                 0.0                   0.0       

In [7]:
test.head()

lot_frontage  lot_area  neighborhood  overall_qual  year_built  \
id                                                                     
2658          69.0      9142             0             6        1910   
2718          68.0      9662             1             5        1977   
2414          58.0     17104             2             7        2006   
1989          60.0      8520             0             5        1923   
625           68.0      9500             1             6        1963   

      year_remod_add  mas_vnr_area  bsmt_qual  bsmt_cond  bsmt_exposure  \
id                                                                        
2658            1950           0.0          2          3              1   
2718            1977           0.0          4          3              1   
2414            2006           0.0          4          4              3   
1989            2006           0.0          3          3              1   
625             1963         247.0          4          3              1   

      bsmtfin_type_1  bsmtfin_sf_1  heating_qc  electrical  1st_flr_sf  \
id                                                                       
2658               1             0           4           1         908   
2718               1             0           3           4        1967   
2414               6           554           5           4         664   
1989               1             0           3           4         968   
625                4           609           4           4        1394   

      2nd_flr_sf  gr_liv_area  bsmt_full_bath  full_bath  half_bath  \
id                                                                    
2658        1020         1928               0          2          0   
2718           0         1967               0          2          0   
2414         832         1496               1          2          1   
1989           0          968               0          1          0   
625            0         1394               1          1          1   

      kitchen_qual  totrms_abvgrd  fireplaces  garage_yr_blt  garage_finish  \
id                                                                            
2658             2              9           0         1910.0              1   
2718             3             10           0         1977.0              3   
2414             4              7           1         2006.0              2   
1989             3              5           0         1935.0              1   
625              3              6           2         1963.0              2   

      garage_cars  garage_area  garage_cond  paved_drive  wood_deck_sf  \
id                                                                       
2658            1          440            1            2             0   
2718            2          580            3            2           170   
2414            2          426            3            2           100   
1989            2          480            3            0             0   
625             2          514            3            2             0   

      open_porch_sf  mas_vnr_type_BrkFace  mas_vnr_type_CBlock  \
id                                                               
2658             60                   0.0                  0.0   
2718              0                   0.0                  0.0   
2414             24                   0.0                  0.0   
1989              0                   0.0                  0.0   
625              76                   1.0                  0.0   

      mas_vnr_type_None  mas_vnr_type_Stone  exterior_2nd_AsphShn  \
id                                                                  
2658                1.0                 0.0                   0.0   
2718                1.0                 0.0                   0.0   
2414                1.0                 0.0                   0.0   
1989                1.0                 0.0                   0.0   
625     

# Preprocessing

## Feature Engineering

Notice that we have many features such full_bath, half_bath and bsmt_full_bath, we can add them altogether to form 1 feature called total_baths

In [8]:
train.shape

(2040, 63)

In [9]:
train['total_baths'] = train['full_bath'] + train['half_bath'] + train['bsmt_full_bath'] 

Since we have a new feature of all the bathrooms together, we can drop the initial seperated bath columns

In [10]:
train.shape

(2040, 64)

In [11]:
train.drop(columns=['full_bath', 'bsmt_full_bath', 'half_bath'], inplace=True)

In [12]:
train.shape

(2040, 61)

We will repeat the same for the test dataset.

In [13]:
test.shape

(879, 62)

In [14]:
test['total_baths'] = test['full_bath'] + test['bsmt_full_bath'] + test['half_bath']

In [15]:
test.drop(columns=['full_bath', 'bsmt_full_bath', 'half_bath'], inplace=True)

In [16]:
test.shape

(879, 60)

## Training Set and Validation Set

Let us break up our train dataset into a training set and a validation set.

In [17]:
X_train = train.drop(columns='saleprice')
y_train = train['saleprice']

In [18]:
X_train.shape

(2040, 60)

In [19]:
y_train.shape

(2040,)

In [20]:
X_train, X_val, y_train, y_val = train_test_split(
    X_train, y_train, test_size=0.25, random_state=100)

In [21]:
X_train.shape

(1530, 60)

In [22]:
X_val.shape

(510, 60)

In [23]:
y_train.shape

(1530,)

In [24]:
y_val.shape

(510,)

We shall assign the test set as X_test and subsequently use this to predict saleprice (y_test) after training our model with the training and validation set.

In [25]:
X_test = test

In [26]:
X_test.shape

(879, 60)

## Scaling

Let's scale our X_train, X_val and X_test data to ensure that our features are uniform in scale.

In [27]:
ss = StandardScaler()

Note that we will would only have to fit the StandardScaler once and transform the data on the X_train, X_val and X_test.

In [28]:
X_train_sc = ss.fit_transform(X_train)
X_val_sc = ss.transform(X_val)
X_test_sc = ss.transform(X_test)

# Modeling

## Linear Regression

In [29]:
# Instantiate the model
lr = LinearRegression()

In [30]:
# Fit the data to the model
lr.fit(X_train_sc, y_train)

LinearRegression()

Our train score without cross validation:

In [31]:
# R^2 score
lr.score(X_train_sc, y_train)

0.8595939244298674

By default, the train score is based on R^2 which is coefficient of determination. It is the proportion of variability in y we can explain with x. Our score means approximately 86% of the saleprice can be explained by our selected features.

We will now do cross validation to compare our scores on the train dataset.

In [32]:
cross_val_score(lr, X_train_sc, y_train, cv=10).mean()

-1.0247430832473963e+22

Linear Regression is not a good model and also not have good at addressing the issue of multicollinearity. Hence, this is shown when putting the model through a more rigorous process of cross val score.

We will also look at cross validating using root mean squared error as our metric.

In [33]:
 np.sqrt(-cross_val_score(lr, X_train_sc, y_train, cv=10, 
                scoring='neg_mean_squared_error').mean())

7946580407044095.0

The model is giving an extremely high error and is severely underfitting as the model is not complex enough. In reality, line of best fit is rarely linear. 

With cross val score, the data is partitioned differently in each fold and it is unable to score well on the all the test sets in each fold, resulting in very high RMSE.

### Evaluating with Root Mean Squared Error (RMSE) on Train and Validation Set

This Kaggle competition calls for the score to be based on RMSE.

We will use X_train_sc to predict saleprice and compare it against y_train data. We will then use RMSE to compare both results. This serves as a guide to how our model predicts saleprice to seen data (train dataset). It is likely that our model will perform slightly poorer to unseen data (test dataset).

In [34]:
np.sqrt(metrics.mean_squared_error(y_train, lr.predict(X_train_sc)))

30251.531728384445

Similarly, we will evaluate our model based on the validation set.

In [35]:
np.sqrt(metrics.mean_squared_error(y_val, lr.predict(X_val_sc)))

32260.01595191512

The goal is to get RMSE to be as close to 0 as possible. Strangely, RMSE is lower in the validation set which shouldn't be the case as training error almost always underestimate validation error.

We will have a look at the saleprice prediction based on the model.

In [36]:
lr_pred = lr.predict(X_test_sc)
lr_pred

array([121029.60415682, 157820.77598682, 197773.51809448,  54538.29907202,
       154889.36292925,  68008.87304339,  74870.71261253, 150334.51262719,
       181387.88050328, 153315.66010671, 144275.03006798, 113682.66948914,
       161234.61475312, 256726.1863586 , 151686.41508511,  60155.7873747 ,
       111371.09561401,  92060.36384589, 190425.11354109, 163432.73226759,
       127768.35158945, 119445.41441483, 170855.29241842, 170319.78051235,
       171300.1976338 , 121338.23196694, 134254.1921901 , 133069.27138333,
       163977.59354692,  30939.12517271,  83075.36539331,  83309.08584055,
       245045.40339924, 170611.39003725, 204730.11332707, 168434.93994176,
        50398.09887627,  39567.67158474, 122346.82529924, 202965.15574617,
       172287.36355943, 199665.32591924, 148777.87259032,  89437.1999191 ,
       236758.19377343,  91247.31222417, 206686.41411522, 116567.66069469,
       101309.05442155, 112478.76797481,  48205.516742  , 163762.22791406,
       254067.93245623,  

## Lasso Regression

In [37]:
# Instantiate the model
lasso_cv = LassoCV(n_alphas=1000, cv=10, max_iter=500000)

In [38]:
# Fit the training data to the model
lasso_cv.fit(X_train_sc, y_train)

LassoCV(cv=10, max_iter=500000, n_alphas=1000)

In [39]:
# Finding the optimal alpha
lasso_cv.alpha_

965.319854180767

This will be our optimal alpha (hyperparamter) which we will put into Lasso model. This optimal alpha penalizes coefficients of features which are too far from zero. As a result, these coefficients are reduced to zeros under Lasso, which is also a form of feature selection.

We will cross validate the lasso R^2 with the optimal alpha.

In [40]:
lasso = Lasso(alpha=lasso_cv.alpha_)

In [41]:
lasso.fit(X_train_sc, y_train)

Lasso(alpha=965.319854180767)

In [42]:
lasso.score(X_train_sc, y_train)

0.8542851576987267

We will look at the lasso coefficients to see which features had the most impact in predicting saleprice.

In [43]:
df_lasso = pd.DataFrame({'features':X_train.columns,
                         'coeff':lasso.coef_})
df_lasso

features         coeff
0           lot_frontage    466.113797
1               lot_area   2148.732760
2           neighborhood  10936.392800
3           overall_qual  18013.147162
4             year_built      0.000000
5         year_remod_add   1179.479642
6           mas_vnr_area   7433.665304
7              bsmt_qual   1917.035933
8              bsmt_cond     -0.000000
9          bsmt_exposure   5807.560500
10        bsmtfin_type_1    358.836904
11          bsmtfin_sf_1   5267.936752
12            heating_qc   2053.640950
13            electrical     -0.000000
14            1st_flr_sf   7947.841686
15            2nd_flr_sf      0.000000
16           gr_liv_area  14996.015650
17          kitchen_qual   9156.877736
18         totrms_abvgrd   3235.119630
19            fireplaces   3964.809642
20         garage_yr_blt  -1424.103068
21         garage_finish    445.679417
22           garage_cars   3772.767705
23           garage_area   4449.075742
24           garage_cond     -0.000000
25           paved_drive      0.000000
26          wood_deck_sf   2147.391125
27         open_porch_sf    649.089131
28  mas_vnr_type_BrkFace  -2341.197729
29     mas_vnr_type_None      0.000000
30    mas_vnr_type_Stone   1630.035288
31  exterior_2nd_AsphShn      0.000000
32  exterior_2nd_Brk Cmn     -0.000000
33  exterior_2nd_BrkFace      0.000000
34   exterior_2nd_CBlock      0.000000
35  exterior_2nd_CmentBd   2762.702628
36  exterior_2nd_HdBoard   -178.961192
37  exterior_2nd_ImStucc     -0.000000
38  exterior_2nd_MetalSd    186.165292
39  exterior_2nd_Plywood     -0.000000
40    exterior_2nd_Stone     -0.000000
41   exterior_2nd_Stucco  -1412.554826
42  exterior_2nd_VinylSd    401.522959
43  exterior_2nd_Wd Sdng     -0.000000
44  exterior_2nd_Wd Shng     -0.000000
45  exterior_1st_AsphShn      0.000000
46  exterior_1st_BrkComm     -0.000000
47  exterior_1st_BrkFace   2278.460313
48   exterior_1st_CBlock      0.000000
49  exterior_1st_CemntBd      0.000000
50  exterior_1st_HdBoard   -549.495115
51  exterior_1st_ImStucc     -0.000000
52  exterior_1st_MetalSd      0.000000
53  exterior_1st_Plywood   -932.657572
54    exterior_1st_Stone     -0.000000
55   exterior_1st_Stucco   -621.441607
56  exterior_1st_VinylSd      0.000000
57  exterior_1st_Wd Sdng     -0.000000
58  exterior_1st_WdShing     -0.000000
59           total_baths   1371.077664

We will also create another column for absolutes of the coefficients to account for highly negative correlation, if any.

In [44]:
df_lasso['absolute_coeff'] = np.abs(df_lasso['coeff'])

In [45]:
df_lasso.sort_values(by='absolute_coeff', ascending=False).head(10)

features         coeff  absolute_coeff
3    overall_qual  18013.147162    18013.147162
16    gr_liv_area  14996.015650    14996.015650
2    neighborhood  10936.392800    10936.392800
17   kitchen_qual   9156.877736     9156.877736
14     1st_flr_sf   7947.841686     7947.841686
6    mas_vnr_area   7433.665304     7433.665304
9   bsmt_exposure   5807.560500     5807.560500
11   bsmtfin_sf_1   5267.936752     5267.936752
23    garage_area   4449.075742     4449.075742
19     fireplaces   3964.809642     3964.809642

Based on Lasso Regression, the 5 most important features as a predictor of saleprice is:
1. overall_qual
2. gr_liv_area
3. neighborhood
4. kitchen_qual
5. 1st_flr_sf

We will also look at the bottom features which represent poor predictors.

We also note that there were features zeroed-out by Lasso. In this case, this is a form of feature selection as well. We will have look at features which were zero-out.

In [46]:
df_lasso.loc[df_lasso['coeff'] == 0, :]

features  coeff  absolute_coeff
4             year_built    0.0             0.0
8              bsmt_cond   -0.0             0.0
13            electrical   -0.0             0.0
15            2nd_flr_sf    0.0             0.0
24           garage_cond   -0.0             0.0
25           paved_drive    0.0             0.0
29     mas_vnr_type_None    0.0             0.0
31  exterior_2nd_AsphShn    0.0             0.0
32  exterior_2nd_Brk Cmn   -0.0             0.0
33  exterior_2nd_BrkFace    0.0             0.0
34   exterior_2nd_CBlock    0.0             0.0
37  exterior_2nd_ImStucc   -0.0             0.0
39  exterior_2nd_Plywood   -0.0             0.0
40    exterior_2nd_Stone   -0.0             0.0
43  exterior_2nd_Wd Sdng   -0.0             0.0
44  exterior_2nd_Wd Shng   -0.0             0.0
45  exterior_1st_AsphShn    0.0             0.0
46  exterior_1st_BrkComm   -0.0             0.0
48   exterior_1st_CBlock    0.0             0.0
49  exterior_1st_CemntBd    0.0             0.0
51  exterior_1st_ImStucc   -0.0             0.0
52  exterior_1st_MetalSd    0.0             0.0
54    exterior_1st_Stone   -0.0             0.0
56  exterior_1st_VinylSd    0.0             0.0
57  exterior_1st_Wd Sdng   -0.0             0.0
58  exterior_1st_WdShing   -0.0             0.0

In [47]:
df_lasso.loc[df_lasso['coeff'] == 0, :].shape

(26, 3)

- There are 12 features which were zero-out by Lasso. 
- These features were computed by the model as features which did not have meaningful coefficients as predictor of the saleprice

We will also assess our model on the validation set.

In [48]:
lasso.score(X_val_sc, y_val)

0.8177922051186162

Lasso's score is marginally lower than Linear Regression's score.

### Evaluating with Root Mean Squared Error (RMSE) on Train and Validation Set

We will use X_train_sc to predict saleprice and compare it against y_train data. We will then use RMSE to compare both results.

In [49]:
np.sqrt(metrics.mean_squared_error(y_train, lasso.predict(X_train_sc)))

30818.13222417267

Similarly, we will evaluate our model based on the validation set.

In [50]:
np.sqrt(metrics.mean_squared_error(y_val, lasso.predict(X_val_sc)))

31915.625737583025

We will have a look at the saleprice prediction based on the model.

In [51]:
lasso_pred = lasso.predict(X_test_sc)
lasso_pred

array([134734.00096761, 167503.97673876, 213290.52741078, 106729.7833669 ,
       161342.13736841,  78549.08540266,  91626.5834034 , 156903.58290718,
       202186.20748335, 162793.13553472, 154835.74623828, 121662.34887977,
       155748.40654493, 265786.66122361, 159559.07806429, 114002.95456663,
       163433.48618592, 106360.01489962, 195389.70074592, 202403.99061216,
       137548.93826249, 129387.51353873, 184055.93166453, 176394.54955276,
       188389.06617464, 130922.99022994, 140264.07253352, 152849.01505188,
       174704.5115497 ,  39121.78392908,  95803.74198503,  98528.33071574,
       232572.5550939 , 174723.22142817, 219328.96730837, 178047.84864846,
       108626.33242182,  88340.46297554, 140105.90800187, 209277.57865008,
       182216.83622242, 213720.59727561, 161333.25141929, 158276.92878714,
       246495.80883678,  92833.33818358, 223085.90529958, 131629.05777699,
       111499.97303642, 118909.66262056, 107209.76688266, 213038.3224246 ,
       260263.20766425, 1

The goal is to get RMSE to be as close to 0 as possible. Compared to our Linear Regression model, Lasso performs more poorly? It could be due to a higher alpha hypermeter which may be 'overpunishing' the features?

## Ridge Regression

In [52]:
# Instantiate the model
ridge_cv = RidgeCV(alphas = np.logspace(0, 5, 100),
                   cv = 10)

In [53]:
# fit the model
ridge_cv.fit(X_train_sc, y_train)

RidgeCV(alphas=array([1.00000000e+00, 1.12332403e+00, 1.26185688e+00, 1.41747416e+00,
       1.59228279e+00, 1.78864953e+00, 2.00923300e+00, 2.25701972e+00,
       2.53536449e+00, 2.84803587e+00, 3.19926714e+00, 3.59381366e+00,
       4.03701726e+00, 4.53487851e+00, 5.09413801e+00, 5.72236766e+00,
       6.42807312e+00, 7.22080902e+00, 8.11130831e+00, 9.11162756e+00,
       1.02353102e+01, 1.14975700e+0...
       6.89261210e+03, 7.74263683e+03, 8.69749003e+03, 9.77009957e+03,
       1.09749877e+04, 1.23284674e+04, 1.38488637e+04, 1.55567614e+04,
       1.74752840e+04, 1.96304065e+04, 2.20513074e+04, 2.47707636e+04,
       2.78255940e+04, 3.12571585e+04, 3.51119173e+04, 3.94420606e+04,
       4.43062146e+04, 4.97702356e+04, 5.59081018e+04, 6.28029144e+04,
       7.05480231e+04, 7.92482898e+04, 8.90215085e+04, 1.00000000e+05]),
        cv=10)

In [54]:
ridge_cv.alpha_

422.9242874389499

This will be our optimal alpha which we will put into Lasso model. This optimal alpha penalizes coefficients of features which are too far from zero, thus enforcing them to small. As a result, these coefficients are reduced to almost zero. This reduces the model complexity while keeping all the variables in the model. 

We will cross validate the ridge R^2 with the optimal alpha.

In [55]:
ridge = Ridge(alpha=ridge_cv.alpha_)

In [56]:
ridge.fit(X_train_sc, y_train)

Ridge(alpha=422.9242874389499)

In [57]:
df_ridge = pd.DataFrame({'features':X_train.columns,
                         'coeff':ridge.coef_})
df_ridge

features         coeff
0           lot_frontage   1841.262877
1               lot_area   2891.937715
2           neighborhood   8687.323258
3           overall_qual  12719.028165
4             year_built    -21.988028
5         year_remod_add   2813.766955
6           mas_vnr_area   7264.888200
7              bsmt_qual   3934.913211
8              bsmt_cond   -525.870830
9          bsmt_exposure   5209.731898
10        bsmtfin_type_1   1708.016962
11          bsmtfin_sf_1   4078.798944
12            heating_qc   2766.358402
13            electrical   -626.700366
14            1st_flr_sf   9189.113527
15            2nd_flr_sf   2039.878159
16           gr_liv_area   8823.436159
17          kitchen_qual   8229.270898
18         totrms_abvgrd   5280.468776
19            fireplaces   5038.462592
20         garage_yr_blt  -1996.406430
21         garage_finish   1851.291592
22           garage_cars   4601.799567
23           garage_area   5117.516787
24           garage_cond   -276.665849
25           paved_drive    600.776258
26          wood_deck_sf   2737.212712
27         open_porch_sf   1836.383749
28  mas_vnr_type_BrkFace  -1374.086383
29     mas_vnr_type_None    612.464053
30    mas_vnr_type_Stone   2802.727821
31  exterior_2nd_AsphShn    151.554744
32  exterior_2nd_Brk Cmn   -211.380090
33  exterior_2nd_BrkFace    645.669308
34   exterior_2nd_CBlock    300.005831
35  exterior_2nd_CmentBd   1839.626566
36  exterior_2nd_HdBoard   -880.420913
37  exterior_2nd_ImStucc   -546.849473
38  exterior_2nd_MetalSd    645.842873
39  exterior_2nd_Plywood   -894.335166
40    exterior_2nd_Stone   -374.121117
41   exterior_2nd_Stucco  -1519.179360
42  exterior_2nd_VinylSd    572.618953
43  exterior_2nd_Wd Sdng    489.881699
44  exterior_2nd_Wd Shng   -605.650178
45  exterior_1st_AsphShn   -242.155876
46  exterior_1st_BrkComm   -117.376209
47  exterior_1st_BrkFace   2493.466746
48   exterior_1st_CBlock   -300.742183
49  exterior_1st_CemntBd   1743.092272
50  exterior_1st_HdBoard   -755.900213
51  exterior_1st_ImStucc    250.969765
52  exterior_1st_MetalSd    399.604889
53  exterior_1st_Plywood   -974.915865
54    exterior_1st_Stone    -56.119087
55   exterior_1st_Stucco   -800.499293
56  exterior_1st_VinylSd    144.330213
57  exterior_1st_Wd Sdng   -808.266904
58  exterior_1st_WdShing     44.105839
59           total_baths   2947.531905

We will also create another column to show absolute coefficients for each features as we also want to see for highly negative correlations which will inversely affect the saleprice.

In [58]:
df_ridge['absolute_coeff'] = np.abs(df_ridge['coeff'])

In [59]:
df_ridge.sort_values(by='absolute_coeff', ascending=False).head(10)

features         coeff  absolute_coeff
3    overall_qual  12719.028165    12719.028165
14     1st_flr_sf   9189.113527     9189.113527
16    gr_liv_area   8823.436159     8823.436159
2    neighborhood   8687.323258     8687.323258
17   kitchen_qual   8229.270898     8229.270898
6    mas_vnr_area   7264.888200     7264.888200
18  totrms_abvgrd   5280.468776     5280.468776
9   bsmt_exposure   5209.731898     5209.731898
23    garage_area   5117.516787     5117.516787
19     fireplaces   5038.462592     5038.462592

Based on Ridge Regression, the 5 most important features as a predictor of saleprice is:
1. overall_qual
2. gr_liv_area
3. neighborhood
4. kitchen_qual
5. 1st_flr_sf

which are the same features for Lasso but interestingly in different orders. 

For example:
- gr_liv_area was ranked 2nd in Lasso but 3rd in Ridge
- neighborhood was ranked 3rd in Lasso but 4th in Ridge
- kitchen_qual was ranked 4th in Lasso but 5th in Ridge
- 1st_flr_sf was ranked 5th in Lasso but 2nd in Ridge

A possible explanation was perhaps how both models treat and penalizes features differently.

We will have a look at features which were poor predictors of saleprice.

In [60]:
df_ridge.sort_values(by='absolute_coeff', ascending=True).head(10)

features       coeff  absolute_coeff
4             year_built  -21.988028       21.988028
58  exterior_1st_WdShing   44.105839       44.105839
54    exterior_1st_Stone  -56.119087       56.119087
46  exterior_1st_BrkComm -117.376209      117.376209
56  exterior_1st_VinylSd  144.330213      144.330213
31  exterior_2nd_AsphShn  151.554744      151.554744
32  exterior_2nd_Brk Cmn -211.380090      211.380090
45  exterior_1st_AsphShn -242.155876      242.155876
51  exterior_1st_ImStucc  250.969765      250.969765
24           garage_cond -276.665849      276.665849

Key observations:
- year_built performed the poorest. This could be explained by referring to the data dictionary. There is also a feature called year remodel, meaning that aged houses could still have high saleprice if it was remodelled recently. We note that this was picked out by the Lasso and zeroed out as a feature.

### Evaluating with Root Mean Squared Error (RMSE) on Train and Validation Set

We will use X_train_sc to predict saleprice and compare it against y_train data. We will then use RMSE to compare both results.

In [61]:
np.sqrt(metrics.mean_squared_error(y_train, ridge.predict(X_train_sc)))

30974.679795240085

Similarly, we will evaluate our model based on the validation set.

In [62]:
np.sqrt(metrics.mean_squared_error(y_val, ridge.predict(X_val_sc)))

32142.312128946047

The goal is to get RMSE to be as close to 0 as possible. 

We will have a look at the saleprice prediction based on the model.

In [63]:
ridge_pred = ridge.predict(X_test_sc)
ridge_pred

array([138339.87842704, 174816.78726824, 216116.34435097,  98694.48661236,
       166527.9758191 ,  79940.70114781,  95913.99997092, 165073.26613065,
       190766.03093732, 166838.64931082, 157146.2915813 , 121486.0137704 ,
       159574.8701486 , 275351.9200119 , 153061.33811024, 104732.13318778,
       156862.4275637 , 108245.3974192 , 202295.10417082, 193210.11402786,
       147072.04517138, 140981.89832361, 185110.99178729, 177986.42376851,
       190118.21041182, 132531.36830967, 146193.00002987, 145753.52673299,
       172988.21898486,  46585.08561004, 105878.16606119, 101303.89100109,
       227416.45747205, 184958.21963473, 217150.30490727, 175949.22647031,
       103944.71271317,  77940.95327221, 141922.64650634, 204107.23540015,
       180264.217663  , 215569.30657742, 165169.68778876, 152808.06305001,
       239702.46149034,  97496.53575963, 219993.58261132, 132039.98998699,
       118861.32898388, 128079.52279608,  96237.20359648, 190137.91819656,
       252960.72521829, 1

## Summary of Model Evaluation

|Model|Hyperparams|No. of Features|No. of Features Columns|Train RMSE|Validation RMSE|
|---|---|---|---|---|---|
|Lasso Regression|optimal alpha = 965|31|60|30,818|31,916|
|Ridge Regression|optimal alpha = 423|31|60|30,975|32,142|
|Linear Regression|N/A|31|60|30,252|32,260|

- Compared to our Linear Regression model, Lasso performs slightly better and the best amongst the 3 models in terms of RMSE
- Our goal is to get RMSE as close to zero as possible 
- Linear Regression has the lowest RMSE on the train set
- Lasso Regression gives the lowest RMSE on the validation set
- We decide to base our choice from the RMSE of the validation set as this set assesses the model's ability to generalize to unseen data
- We will be using Lasso model to predict the saleprice on our test dataset.

# Retrain on the maximum train dataset with chosen model

Once we have chosen our model, we will retrain on the maximal amount of data using the full train set.

In [64]:
X_train_final = train.drop(columns='saleprice')
y_train_final = train['saleprice']

In [65]:
lasso.fit(X_train_final, y_train_final)

Lasso(alpha=965.319854180767)

In [66]:
np.sqrt(metrics.mean_squared_error(y_train_final, lasso.predict(X_train_final)))

31618.09556576091

# Prediction of salepice on test set

In [67]:
lasso_pred

array([134734.00096761, 167503.97673876, 213290.52741078, 106729.7833669 ,
       161342.13736841,  78549.08540266,  91626.5834034 , 156903.58290718,
       202186.20748335, 162793.13553472, 154835.74623828, 121662.34887977,
       155748.40654493, 265786.66122361, 159559.07806429, 114002.95456663,
       163433.48618592, 106360.01489962, 195389.70074592, 202403.99061216,
       137548.93826249, 129387.51353873, 184055.93166453, 176394.54955276,
       188389.06617464, 130922.99022994, 140264.07253352, 152849.01505188,
       174704.5115497 ,  39121.78392908,  95803.74198503,  98528.33071574,
       232572.5550939 , 174723.22142817, 219328.96730837, 178047.84864846,
       108626.33242182,  88340.46297554, 140105.90800187, 209277.57865008,
       182216.83622242, 213720.59727561, 161333.25141929, 158276.92878714,
       246495.80883678,  92833.33818358, 223085.90529958, 131629.05777699,
       111499.97303642, 118909.66262056, 107209.76688266, 213038.3224246 ,
       260263.20766425, 1

In [68]:
test['SalePrice'] = lasso_pred
test.head()

lot_frontage  lot_area  neighborhood  overall_qual  year_built  \
id                                                                     
2658          69.0      9142             0             6        1910   
2718          68.0      9662             1             5        1977   
2414          58.0     17104             2             7        2006   
1989          60.0      8520             0             5        1923   
625           68.0      9500             1             6        1963   

      year_remod_add  mas_vnr_area  bsmt_qual  bsmt_cond  bsmt_exposure  \
id                                                                        
2658            1950           0.0          2          3              1   
2718            1977           0.0          4          3              1   
2414            2006           0.0          4          4              3   
1989            2006           0.0          3          3              1   
625             1963         247.0          4          3              1   

      bsmtfin_type_1  bsmtfin_sf_1  heating_qc  electrical  1st_flr_sf  \
id                                                                       
2658               1             0           4           1         908   
2718               1             0           3           4        1967   
2414               6           554           5           4         664   
1989               1             0           3           4         968   
625                4           609           4           4        1394   

      2nd_flr_sf  gr_liv_area  kitchen_qual  totrms_abvgrd  fireplaces  \
id                                                                       
2658        1020         1928             2              9           0   
2718           0         1967             3             10           0   
2414         832         1496             4              7           1   
1989           0          968             3              5           0   
625            0         1394             3              6           2   

      garage_yr_blt  garage_finish  garage_cars  garage_area  garage_cond  \
id                                                                          
2658         1910.0              1            1          440            1   
2718         1977.0              3            2          580            3   
2414         2006.0              2            2          426            3   
1989         1935.0              1            2          480            3   
625          1963.0              2            2          514            3   

      paved_drive  wood_deck_sf  open_porch_sf  mas_vnr_type_BrkFace  \
id                                                                     
2658            2             0             60                   0.0   
2718            2           170              0                   0.0   
2414            2           100             24                   0.0   
1989            0             0              0                   0.0   
625             2             0             76                   1.0   

      mas_vnr_type_CBlock  mas_vnr_type_None  mas_vnr_type_Stone  \
id                                                                 
2658                  0.0                1.0                 0.0   
2718                  0.0                1.0                 0.0   
2414                  0.0                1.0                 0.0   
1989                  0.0                1.0                 0.0   
625                   0.0                0.0                 0.0   

      exterior_2nd_AsphShn  exterior_2nd_Brk Cmn  exterior_2nd_BrkFace  \
id                                                                       
2658                   0.0                   0.0                   0.0   
2718                   0.0                   0.0                   0.0   
2414                   0.0                   0.0                   0.0   
1989                   0.0                   

In [69]:
submission = test[['SalePrice']]
submission.index.rename('Id', inplace=True)
submission.sort_index(inplace=True)
submission

C:\Users\caspe\anaconda3\envs\firstday\lib\site-packages\pandas\core\frame.py:5582: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().sort_index(


SalePrice
Id                 
2     118011.999957
4     264840.553941
6     189307.943633
7     231360.203675
17    223321.367556
18    337196.140344
22    196955.752553
27     95636.185054
31    105179.545066
36    156286.627336
39    346065.833076
40    250864.903608
42    305812.930747
46    231183.559400
53    190036.220707
58    182453.730659
61    283511.847215
72    229650.539832
79    112647.666218
80    210815.868534
82    160049.835501
83    118218.909699
85    168822.216887
88    118909.662621
92    289329.586829
93    281323.720789
95    202186.207483
102   228476.288069
103   202340.093356
106   251927.236083
107   195990.427682
110   239770.067276
114   204859.287526
116   179708.717537
117   164921.174598
118   138712.095505
125   174704.511550
126   167728.593064
127   140384.911828
128   204864.082145
134   135154.925739
135   181342.251697
137   239956.866931
143   137411.195586
147   121034.372314
148   197886.429618
150    97317.818815
156   124754.589508
157   111499.973036
159   127652.301080
161   269711.500589
167   137756.235759
170   156547.920454
171    89636.294009
175    94681.779437
179   106486.237586
181   110327.759578
186   213038.322425
193   203191.346808
194    94621.162523
206   121429.460149
208   174074.675741
210   185928.611403
211   100132.540899
212   130888.720830
214   117980.497858
215   162421.549473
217   126298.675050
222   137548.938262
227   152123.859512
228   168796.410439
233   228675.685203
235   110840.543427
238   134441.847495
241   164591.997811
252   204202.378959
253   193740.537513
256   160066.021345
257   159090.984522
258   203028.984843
259   190121.896009
261   152113.238749
263   141999.784358
266   226397.575050
267   237751.491544
270   170191.767310
271   253976.119409
276   134245.422462
285   160185.419653
288    60868.273619
289   150319.474237
290   125328.899627
291   143770.031509
292   134626.257960
293   217438.240141
294   224543.838244
295   202232.770998
304   122579.389277
322   313509.093063
326   169846.817716
327    85949.287263
328   134307.569886
331    74742.650436
332    74206.238231
333    78549.085403
335   138903.779981
337   195498.330957
342   121669.145856
343   167656.483304
347   178836.733461
350   327507.118618
352   227506.433424
358   159712.057819
359   185655.825624
360   164569.673086
368   304967.004497
369   253975.595466
373   177148.120941
377   237500.415288
383   176471.649063
386   241729.096424
387   243887.953631
390   157340.739086
391   163728.524464
395   140377.735461
397   107534.524494
398   110178.619066
401   107971.734084
404   123620.691849
408   105173.129159
409   131201.922436
410   161885.588819
418   134588.955490
422   387283.783722
427   309519.089689
431   344404.374687
433   423872.576021
437   329582.492694
438   281293.130103
442   312555.498594
447   226582.666515
459   301467.095822
460   272708.693890
464   206020.339663
465   193823.493128
467   183970.554719
468   192406.833083
474   181399.494921
480   208186.947347
481   204715.976625
483   229288.735462
488   172186.683630
492   180948.574972
494   173622.869674
499   323933.943706
500   289440.815647
501   316207.747949
503   254614.685007
506   240826.693576
508   266825.519117
511   291550.491311
515   241764.606383
519   235603.849353
521   281331.926779
530   239969.821051
532   233780.768406
534   188389.066175
540   186760.942197
541   141015.123989
551   148649.721910
552   171627.937981
554   118164.929244
555   104304.113492
556   105597.362907
557   129387.513539
565   236355.221171
566   279210.358974
567   182216.836222
569   229585.427324
570   215560.050052
574   243784.591082
575   310220.581079
576   217790.593390
577   237459.696289
578   227654.651078
583   236513.240452
584   244063.598425
585   154211.083766
587   209171.679924
593   214203.007183
597   194674.541044
598   194734.247216
599   166353.651710
600   112341.933682
605   171327.410886
607   165540.093485
608   160336.134921
611   185314.168098
616   1352

In [70]:
submission.shape

(879, 1)

In [71]:
submission.to_csv('../datasets/submission.csv')